src/main/java/com/mraof/minestuck/item/StructureScannerItem.java

```diff
@@ -2,84 +2,180 @@
+	 * Set the location to the nearest structure, check that a structure exists, then reduce charge if fuelled.
+	 */
+	@Override
+	public void inventoryTick(ItemStack pStack, Level pLevel, Entity pEntity, int pSlotId, boolean pIsSelected)
+	{
+		super.inventoryTick(pStack, pLevel, pEntity, pSlotId, pIsSelected);
+		
+		if(pStack.hasTag() && pStack.getTag().getBoolean("Powered") && pLevel instanceof ServerLevel sLevel)
+		{
+			if (!isCharged(pStack))
+			{
+				resetCharge(pStack);
 			}
+			
+			BlockPos pos = setLocation(pEntity, sLevel);
+			
+			checkLocation(pStack, pos);
+			
+			reduceCharge(pStack, pEntity, pLevel);
+		}
+	}
+	
+	public static boolean isCharged(ItemStack stack)
+	{
+		return stack.getDamageValue() < stack.getMaxDamage();
+	}
+	
+	public void resetCharge(ItemStack pStack)
+	{
+		pStack.setDamageValue(0);
+	}
+	
+	public BlockPos setLocation(Entity pEntity, ServerLevel sLevel)
+	{
+		return sLevel.findNearestMapStructure(structure, pEntity.blockPosition(), 100, false);
+	}
+	
+	public void checkLocation(ItemStack pStack, BlockPos pos)
+	{
+		if(pos == null)
+		{
+			pStack.getTag().remove("TargetLocation");
 		} else
 		{
-			foundItem = true;
+			pStack.getTag().put("TargetLocation", NbtUtils.writeBlockPos(pos));
 		}
-		
-		if(levelIn instanceof ServerLevel level)
+	}
+	
+	/**
+	 * Check that the fuel item is set; if not, don't reduce charge.
+	 * Scanner charge is represented by durability and damage. Damage is dealt every 20 ticks.
+	 * If it runs out of charge, power it off.
+	 *
+	 * @param pStack current item
+	 * @param pEntity player entity
+	 * @param pLevel player's current level
+	 * @return Boolean check for if deactivated, used to reactivate if players still has fuel.
+	 */
+	public void reduceCharge(ItemStack pStack, Entity pEntity, Level pLevel)
+	{
+		if(fuelItem != null && pEntity.tickCount % 20 == 0)
 		{
-			if(foundItem || playerIn.isCreative())
+			pStack.hurt(1, pLevel.random, pEntity instanceof ServerPlayer ? (ServerPlayer) pEntity : null);
+			
+			if(!isCharged(pStack))
 			{
-				BlockPos structureBlockPos = level.findNearestMapStructure(structure, playerIn.blockPosition(), 100, false);
+				pStack.getTag().putBoolean("Powered", false);
 				
-				if(structureBlockPos != null)
-				{
-					BlockPos playerBlockPos = playerIn.blockPosition();
-					playerIn.sendSystemMessage(Component.translatable(getDescriptionId() + ".successMessage",
-							(int) Math.sqrt(playerBlockPos.distSqr(structureBlockPos.above(64)))).withStyle(ChatFormatting.AQUA));
-				} else
-				{
-					playerIn.sendSystemMessage(Component.translatable(getDescriptionId() + ".failMessage").withStyle(ChatFormatting.RED));
+				if (!pLevel.isClientSide()){
+					MutableComponent message = Component.translatable("message.temple_scanner.off");
+					pEntity.sendSystemMessage(message.withStyle(ChatFormatting.DARK_GREEN));
 				}
-			} else
+			}
+		}
+	}
+	
+	public boolean checkFuelNeeded(Player pPlayer, Level pLevel)
```

Comment:
This function might need some form of renaming.
Also, if I interpret the code correctly, it currently prevents a scanner item with `fuelItem == null` from being turned on. Does that seem right? If so, that needs fixing if we're gonna maintain that `fuelItem` can be null.

Reply:
Resolved; now it returns true if it's null, and only returns false when it isn't null and the player's inventory doesn't have any fuel in it.

Renamed functions to the following:
- hasFuel => findFuel
- checkFuelNeeded => getFuel
- useFuelItem => useFuel

Reply:
I'd only argue against the "checkFuelNeeded => getFuel" rename. I do have an idea though. If you're fine with it, I could implement it myself and commit it directly to the PR.
My idea is to first convert the `if(getFuel(pPlayer, pLevel) || pPlayer.isCreative())` check in `use()` into an early return, and to then try inlining the `getFuel()` function into `use()`.

Are you fine with me doing this?